Wikipedia/Code/WMFArticleImageProtocol.m

```diff
@@ -55,12 +71,44 @@ - (void)startLoading {
     DDLogVerbose(@"Fetching image %@", self.request.URL);
     @weakify(self);
     [[WMFImageController sharedInstance] fetchImageWithURL:self.request.URL]
-    .then(^(WMFImageDownload* download) {
+    .thenInBackground(^WMFArticleImageProtocolResponseAndData*(WMFImageDownload* download) {
         @strongify(self);
         if(!self){
-            return;
+            return nil;
         }
-        [self respondWithDataFromDownload:download];
+        return [self getResponseAndDataForDownload:download];
+    })
+    .then(^id(WMFArticleImageProtocolResponseAndData* responseAndData) {
+        @strongify(self);
+        if(!self){
+            return nil;
+        }
+
+        if (!responseAndData.response) {
```

Comment:
If there were any problems making a response or processing the image (on the background thread) we now return errors so the catch block will fire. Formerly if anything went wrong in either of these parts "URLProtocolDidFinishLoading" was still invoked, which indicated success, when in fact the loading had failed, which is what the catch block correctly reports.

I very much suspect by not correctly reporting failure there were protocol bits behind the scenes which were then hanging around too long any time there were issues retrieving images.